sql: enable adding columns which are not written to the current primary index

postamar opened the pull request:
row,schemachanger: use StoreColumnIDs/Names in primary index
    
    In a recent commit, the StoreColumnIDs and StoreColumnNames slices in
    primary indexes were populated when previously they had simply been
    empty. We simply assumed that all non-virtual columns in a table would
    be stored in the primary index: primary key columns in the key, the rest
    in the value.
    
    This commit breaks that assumption by using the StoreColumnIDs slice to
    determine what goes into the primary index. This makes it possible for
    the new schema changer to add columns safely, preventing unwanted writes
    to the old primary index while the schema change is underway.
    
    Fixes #59149.
    
    Release note: None


    sql,tabledesc: add new IndexDescriptorVersion for primary indexes
    
    Previously, the IndexDescriptorVersion type was only used to describe
    the encoding of secondary indexes. This commit adds a new value for
    use in primary indexes, PrimaryIndexWithStoredColumnsVersion, to signify
    that the StoredColumnIDs and StoredColumnNames slices are populated
    correctly.
    
    Previously, these slices did not need to be populated at all. This is
    because the set of columns comprising the primary index of a table is
    assumed to be all non-virtual columns of that table. Our upcoming work on
    the new schema changer will require us to violate that assumption
    however. This commit is in preparation of that change.
    
    Release note: None


    backupccl: run post-deserialization upgrades on restored descriptors
    
    Previously, we ran these upgrades only for the purpose of upgrading
    foreign keys in tables. This appears to be an omission, the full suite
    of upgrades should be run after a restore. The reason for this is that
    a descriptor which was valid in an old cluster version may no longer be
    valid (according to the descriptor validation checks) in a more recent
    cluster version. For example, a table descriptor has a format version
    field; its value may correctly be `BaseFormatVersion` in an old cluster
    but recent clusters expect at least a `FamilyFormatVersion`.
    
    Release note (bug fix): Fixed a bug which would have prevented old
    backups from being successfully restored.

Commit:
backupccl: run post-deserialization upgrades on restored descriptors

Previously, we ran these upgrades only for the purpose of upgrading
foreign keys in tables. This appears to be an omission, the full suite
of upgrades should be run after a restore. The reason for this is that
a descriptor which was valid in an old cluster version may no longer be
valid (according to the descriptor validation checks) in a more recent
cluster version. For example, a table descriptor has a format version
field; its value may correctly be `BaseFormatVersion` in an old cluster
but recent clusters expect at least a `FamilyFormatVersion`.

Release note (bug fix): Fixed a bug which would have prevented old
backups from being successfully restored.

Commit:
sql,tabledesc: add new IndexDescriptorVersion for primary indexes

Previously, the IndexDescriptorVersion type was only used to describe
the encoding of secondary indexes. This commit adds a new value for
use in primary indexes, PrimaryIndexWithStoredColumnsVersion, to signify
that the StoredColumnIDs and StoredColumnNames slices are populated
correctly.

Previously, these slices did not need to be populated at all. This is
because the set of columns comprising the primary index of a table is
assumed to be all non-virtual columns of that table. Our upcoming work on
the new schema changer will require us to violate that assumption
however. This commit is in preparation of that change.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/66494)
<!-- Reviewable:end -->

Commit:
row,schemachanger: use StoreColumnIDs/Names in primary index

In a recent commit, the StoreColumnIDs and StoreColumnNames slices in
primary indexes were populated when previously they had simply been
empty. We simply assumed that all non-virtual columns in a table would
be stored in the primary index: primary key columns in the key, the rest
in the value.

This commit breaks that assumption by using the StoreColumnIDs slice to
determine what goes into the primary index. This makes it possible for
the new schema changer to add columns safely, preventing unwanted writes
to the old primary index while the schema change is underway.

Fixes #59149.

Release note: None

postamar commented:
@ajwerner I'd appreciate it if you could please give this a quick 👀 before I refine this, I need a sanity check at this point.

 If I'm on the right track, I'll supplement this with more tests. Some areas of this change are dangerously under-tested:
- Anything to do with partitions. I must have changed the contents of `tabledesc.UpdateIndexPartitioning` three times and never did a test case fail as a result.
- The first commit in the PR I suspect fixes a long-standing bug. I should be able to come up with a test case that proves this, I just haven't gotten around to it. If my hunch is correct, this commit should be its own PR so as to backport it.
- We can probably afford to be stricter with `version` and `encoding_type` in `IndexDescriptor` in validation checks.
- etc.

ajwerner commented on pkg/sql/catalog/tabledesc/table_desc.go:
```diff
@@ -183,6 +184,33 @@ func UpdateIndexPartitioning(
+		id := idx.StoreColumnIDs[i]
+		name := idx.StoreColumnNames[i]
+		found := false
+		for _, newColumnName := range newColumnNames {
```
why do this search by name and not by ID? I don't recall how `ALTER COLUMN TYPE` works exactly but I recall a new column being added with the same name and then being swapped it. I don't remember if we use a temp name.

postamar replied:
I can't remember but it might have been the case that I thought (mistakenly or not) that some IDs might still be zero at this point. I hope I'm wrong. This just reinforces my feeling that this stuff is undertested.

postamar replied:
Confirming that this needs to be by name because this can be called pre-ID-allocation.

ajwerner commented on pkg/sql/schemachanger/schemachanger_test.go:
```diff
@@ -412,3 +412,120 @@ func TestConcurrentOldSchemaChangesCannotStart(t *testing.T) {
+
+	<-beforeBackfillNotification
+
+	conn, err := sqlDB.Conn(ctx)
```
Why the `Conn` here? Also,` sqlutils.MakeSQLRunner` is your friend

postamar replied:
Oh I just copy/pasted from the other test in the same file. I'll improve both.

ajwerner reviewed (commented):
Seems like a fine approach to me. We probably want to get rid of the behavior change regarding the `ALTER PRIMARY KEY` and the created unique index.

postamar commented:
Thanks for the reassurance Andrew! Much appreciated.

Commit:
fixup schemachanger_test.go

Commit:
fixup UpdateIndexPartitioning
